opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -11,3 +11,365 @@
+            aggregation: The type of aggregation to use for this measure metric.
+            span_context: The `SpanContext` that identified the `Span`
+                for which the metrics are associated with.
+
+        Returns:
+            A new `MeasureFloat`
+        """
+
+
+class Metric(ABC):
+    """Base class for various types of metrics.
+
+    Metric class that inherit from this class are specialized with the type of
+    time series that the metric holds. Metric is constructed from the meter.
+    """
+
+    @abstractmethod
+    def get_or_create_time_series(self,
+                                  label_values: LabelValues
+                                  ) -> 'object':
+        """Gets and returns a timeseries, a container for a cumulative value.
+
+        If the provided label values are not already associated with this
+        metric, a new timeseries is returned, otherwise it returns the existing
+        timeseries with the exact label values. The timeseries returned
+        contains logic and behaviour specific to the type of metric that
+        overrides this function.
+
+        Args:
+            label_values: A map of `LabelValue` s that will be associated
+                with the return timeseries.
+        """
+
+    @abstractmethod
+    def get_default_time_series(self) -> 'object':
+        """Returns a timeseries, a container for a cumulative value.
+
+        The timeseries will have all its labels not set (default).
+        """
+
+    def set_call_back(self,
+                      updater_function: Callable[..., None]
+                      ) -> None:
+        """Sets a callback that gets executed every time prior to exporting.
+
+        This function MUST set the value of the `Metric` to the value that
+        that will be exported.
+
+        args:
+            updater_function: The callback function to execute.
+        """
+
+    def remove_time_series(self,
+                           label_values: LabelValues) -> None:
+        """Removes the timeseries from the Metric, if present.
+
+        The timeseries with matching `LabelValue` s will be removed.
+
+        args:
+            label_values: The list of label values to match against.
+        """
+
+    def clear(self) -> None:
+        """Removes all timeseries from the `Metric`."""
+
+
+class CounterFloat(Metric):
+    """A counter type metric that holds float values.
+
+    Cumulative values can go up or stay the same, but can never go down.
+    Cumulative values cannot be negative.
+    """
+
+    def get_or_create_time_series(self,
+                                  label_values: LabelValues
+                                  ) -> 'CounterTimeSeries':
+        """Gets a `CounterTimeSeries` with a cumulative float value."""
+
+    def get_default_time_series(self) -> 'CounterTimeSeries':
+        """Returns a `CounterTimeSeries` with a cumulative float value."""
+
+
+class CounterInt(Metric):
```

Comment:
We see this problem with almost every class in the metrics package. Tracing has separate API and SDK packages largely because vendors want to be able to change the implementation. Metrics has separate API and SDK packages to be consistent with tracing, but if (e.g.) vendors are going to change metrics, they're much more likely to add new aggregations or metric types than to change these built-in classes.

This touches on a deeper issue regarding the API/SDK separation. The API already includes some implementation to support context propagation without an SDK, so how do we decide where to draw the line?

Reply:
Sounds like a great question to discuss in the specification. I agree with all the points you've raised here.

Reply:
The current metrics RFC discusses the semantic connection between metrics APIs and trace context. The current specification describes three kinds of metric "instrument", cumulative, gauge, and measure, which are pure API objects. Handles (known as `TimeSeries` currently) are SDK-specific implementations of the `Add()`, `Set()`, and `Record()` APIs. Nothing is said about aggregations at the API level apart from the the standard interpretation given to each, which is to say that by default cumulative metric instruments aggregate a sum, gauge metric instruments aggregate a last-value, and measure instruments aggregate a rate of some distribution.
https://github.com/open-telemetry/oteps/pull/29

By separating the API from the implementation, it should be possible for to define an exporter for direct integration with existing metrics client libraries. This will require some sort of new metrics event exporter to be specified.